module/state_synchronization/storage.go

```diff
@@ -0,0 +1,957 @@
+			if err != nil {
+				return fmt.Errorf("failed to retrieve blob record value for Cid %s: %w", c.String(), err)
+			}
+
+			latestHeightContainingBlob := binary.LittleEndian.Uint64(blobRecordValue)
+			if latestHeightContainingBlob < blockHeight {
+				// this should never happen
+				return fmt.Errorf(
+					"inconsistency detected: blob record height for Cid %s is %d, but blob tree record exists at height %d",
+					c.String(), latestHeightContainingBlob, blockHeight,
+				)
+			} else if latestHeightContainingBlob == blockHeight {
+				if err := txn.Delete(blobRecordKey); err != nil {
+					return fmt.Errorf("failed to delete blob record for Cid %s: %w", c.String(), err)
+				}
+
+				blobKey := network.BlobServiceDatastoreKeyFromCid(c).Bytes()
+				if err := txn.Delete(blobKey); err != nil {
+					return fmt.Errorf("failed to delete blob %s: %w", c.String(), err)
+				}
+
+				deletedCIDs[n] = c
+				n += 1
+			}
+		}
+
+		numDeleted = n
+
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	return numDeleted, nil
+}
+
+func (s *Storage) getCidsToDelete(height uint64) ([]cid.Cid, []uint64, error) {
+	var cidsToDelete []cid.Cid
+	var blockHeights []uint64
+
+	blobTreeRecordPrefix := []byte{prefixBlobTreeRecord}
+
+	if err := s.db.View(func(txn *badger.Txn) error {
+		it := txn.NewIterator(badger.IteratorOptions{
+			PrefetchValues: false,
+			Prefix:         blobTreeRecordPrefix,
+		})
+		defer it.Close()
+
+		for it.Seek(blobTreeRecordPrefix); it.ValidForPrefix(blobTreeRecordPrefix); it.Next() {
+			item := it.Item()
+			blobTreeRecordKey := item.Key()
+
+			blockHeight, blobCid, err := parseBlobTreeRecordKey(blobTreeRecordKey)
+			if err != nil {
+				return fmt.Errorf("malformed blob tree record key %v: %w", blobTreeRecordKey, err)
+			}
+
+			if blockHeight > height {
+				break
+			}
+
+			cidsToDelete = append(cidsToDelete, blobCid)
+			blockHeights = append(blockHeights, blockHeight)
+		}
+
+		return nil
+	}); err != nil {
+		return nil, nil, err
+	}
+
+	return cidsToDelete, blockHeights, nil
+}
+
+func (s *Storage) init(startHeight uint64) error {
+	storedDataLowerBound, err1 := s.GetStoredDataLowerBound()
+	latestIncorporatedHeight, err2 := s.GetLatestIncorporatedHeight()
+
+	if err1 == nil && err2 == nil {
+		if err := s.replay(storedDataLowerBound, latestIncorporatedHeight); err != nil {
+			return fmt.Errorf("failed to restore db: %w", err)
+		}
+	} else if errors.Is(err1, badger.ErrKeyNotFound) && errors.Is(err2, badger.ErrKeyNotFound) {
+		if err := s.bootstrap(startHeight); err != nil {
+			return fmt.Errorf("failed to bootstrap db: %w", err)
+		}
+	} else if (errors.Is(err1, badger.ErrKeyNotFound) && err2 == nil) || (err1 == nil && errors.Is(err2, badger.ErrKeyNotFound)) {
+		return fmt.Errorf("db is in inconsistent state")
+	} else {
+		return fmt.Errorf("failed to initialize storage: %w", err1)
+	}
+
+	return nil
+}
+
+func (s *Storage) replay(storedDataLowerBound uint64, latestIncorporatedHeight uint64) error {
```

Comment:
Is there any assumption that storedDataLowerBound must be lower than latestIncorporatedHeight? Do we need any sanity check here?

Reply:
It definitely should be less than or equal to the latestIncorporatedHeight. This is checked by the `Check` function. But I think you're right that we should check this here as well, since this is called during the actual creation of the storage.